File: DynamicAnalyzer/tools/adb/linux/systrace/catapult/telemetry/telemetry/internal/actions/pinch.js
// Copyright 2013 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

// This file provides the PinchAction object, which zooms into or out of a
// page by a given scale factor:
//   1. var action = new __PinchAction(callback)
//   2. action.start(pinch_options)
'use strict';

(function() {

  function PinchGestureOptions(opt_options) {
    if (opt_options) {
      this.element_ = opt_options.element;
      this.left_anchor_ratio_ = opt_options.left_anchor_ratio;
      this.top_anchor_ratio_ = opt_options.top_anchor_ratio;
      this.scale_factor_ = opt_options.scale_factor;
      this.speed_ = opt_options.speed;
    } else {
      this.element_ = document.body;
      this.left_anchor_ratio_ = 0.5;
      this.top_anchor_ratio_ = 0.5;
      this.scale_factor_ = 2.0;
      this.speed_ = 800;
    }
  }

  function supportedByBrowser() {
    return !!(window.chrome &&
              chrome.gpuBenchmarking &&
              chrome.gpuBenchmarking.pinchBy &&
              chrome.gpuBenchmarking.visualViewportHeight &&
              chrome.gpuBenchmarking.visualViewportWidth);
  }

  // This class zooms into or out of a page, given a number of pixels for
  // the synthetic pinch gesture to cover.
  function PinchAction(opt_callback) {
    var self = this;

    this.beginMeasuringHook = function() {};
    this.endMeasuringHook = function() {};

    this.callback_ = opt_callback;
  };

  PinchAction.prototype.start = function(opt_options) {
    this.options_ = new PinchGestureOptions(opt_options);

    requestAnimationFrame(this.startPass_.bind(this));
  };

  PinchAction.prototype.startPass_ = function() {
    this.beginMeasuringHook();

    var rect = __GestureCommon_GetBoundingVisibleRect(this.options_.element_);
    var anchor_left =
        rect.left + rect.width * this.options_.left_anchor_ratio_;
    var anchor_top =
        rect.top + rect.height * this.options_.top_anchor_ratio_;
    chrome.gpuBenchmarking.pinchBy(this.options_.scale_factor_,
                                   anchor_left, anchor_top,
                                   this.onGestureComplete_.bind(this),
                                   this.options_.speed_);
  };

  PinchAction.prototype.onGestureComplete_ = function() {
    this.endMeasuringHook();

    if (this.callback_)
      this.callback_();
  };

  window.__PinchAction = PinchAction;
  window.__PinchAction_SupportedByBrowser = supportedByBrowser;
})();

File: DynamicAnalyzer/tools/adb/windows/systrace/catapult/telemetry/telemetry/internal/actions/pinch.py
# Copyright 2013 The Chromium Authors. All rights reserved.
# Use of this source code is governed by a BSD-style license that can be
# found in the LICENSE file.

from telemetry.internal.actions import page_action
from telemetry.internal.actions import utils
from telemetry.util import js_template


class PinchAction(page_action.PageAction):
  def __init__(self, selector=None, text=None, element_function=None,
               left_anchor_ratio=0.5, top_anchor_ratio=0.5,
               scale_factor=None, speed_in_pixels_per_second=800,
               synthetic_gesture_source=page_action.GESTURE_SOURCE_DEFAULT):
    super(PinchAction, self).__init__()
    self._selector = selector
    self._text = text
    self._element_function = element_function
    self._left_anchor_ratio = left_anchor_ratio
    self._top_anchor_ratio = top_anchor_ratio
    self._scale_factor = scale_factor
    self._speed = speed_in_pixels_per_second
    self._synthetic_gesture_source = ('chrome.gpuBenchmarking.%s_INPUT' %
                                      synthetic_gesture_source)

    if (self._selector is None and self._text is None and
        self._element_function is None):
      self._element_function = 'document.body'

  def WillRunAction(self, tab):
    utils.InjectJavaScript(tab, 'gesture_common.js')
    utils.InjectJavaScript(tab, 'pinch.js')

    # Fail if browser doesn't support synthetic pinch gestures.
    if not tab.EvaluateJavaScript('window.__PinchAction_SupportedByBrowser()'):
      raise page_action.PageActionNotSupported(
          'Synthetic pinch not supported for this browser')

    tab.ExecuteJavaScript("""
        window.__pinchActionDone = false;
        window.__pinchAction = new __PinchAction(function() {
          window.__pinchActionDone = true;
        });""")

  @staticmethod
  def _GetDefaultScaleFactorForPage(tab):
    current_scale_factor = tab.EvaluateJavaScript(
        'window.outerWidth / window.innerWidth')
    return 3.0 / current_scale_factor

  def RunAction(self, tab):
    scale_factor = (self._scale_factor if self._scale_factor else
                    PinchAction._GetDefaultScaleFactorForPage(tab))
    code = js_template.Render('''
        function(element, info) {
          if (!element) {
            throw Error('Cannot find element: ' + info);
          }
          window.__pinchAction.start({
            element: element,
            left_anchor_ratio: {{ left_anchor_ratio }},
            top_anchor_ratio: {{ top_anchor_ratio }},
            scale_factor: {{ scale_factor }},
            speed: {{ speed }}
          });
        }''',
        left_anchor_ratio=self._left_anchor_ratio,
        top_anchor_ratio=self._top_anchor_ratio,
        scale_factor=scale_factor,
        speed=self._speed)
    page_action.EvaluateCallbackWithElement(
        tab, code, selector=self._selector, text=self._text,
        element_function=self._element_function)
    tab.WaitForJavaScriptCondition('window.__pinchActionDone', timeout=60)
